flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/MapUnionFunction.java

```diff
@@ -0,0 +1,161 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.GenericArrayData;
+import org.apache.flink.table.data.MapData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.FunctionContext;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.KeyValueDataType;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+import java.lang.invoke.MethodHandle;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.flink.table.api.Expressions.$;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#MAP_UNION}. */
+@Internal
+public class MapUnionFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter keyElementGetter;
+    private final ArrayData.ElementGetter valueElementGetter;
+
+    private final SpecializedFunction.ExpressionEvaluator keyEqualityEvaluator;
+    private transient MethodHandle keyEqualityHandle;
+
+    public MapUnionFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.MAP_UNION, context);
+        KeyValueDataType outputType =
+                ((KeyValueDataType) context.getCallContext().getOutputDataType().get());
+        final DataType keyDataType = outputType.getKeyDataType();
+        final DataType valueDataType = outputType.getValueDataType();
+        keyElementGetter =
+                ArrayData.createElementGetter(outputType.getKeyDataType().getLogicalType());
+        valueElementGetter =
+                ArrayData.createElementGetter(outputType.getValueDataType().getLogicalType());
+        keyEqualityEvaluator =
+                context.createEvaluator(
+                        $("element1").isEqual($("element2")),
+                        DataTypes.BOOLEAN(),
+                        DataTypes.FIELD("element1", keyDataType.notNull().toInternal()),
+                        DataTypes.FIELD("element2", keyDataType.notNull().toInternal()));
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        keyEqualityHandle = keyEqualityEvaluator.open(context);
+    }
+
+    public @Nullable MapData eval(@Nullable MapData... maps) {
+        try {
+            if (maps == null || maps.length == 0) {
+                return null;
+            }
+            if (maps.length == 1) {
+                return maps[0];
+            }
+            MapData result = maps[0];
+            for (int i = 1; i < maps.length; ++i) {
+                MapData map = maps[i];
+                if (map != null && map.size() > 0) {
```

Comment:
I wonder whether we need this check?
If in doc we mention that it should return null if any of maps is null then IMHO once we encountered the first null map we don't need to check the rest and can immediatly return null. Or did I miss anything?

Reply:
Yes, if map == null, we should return null.